### PR TITLE
Feature/additional merge pass after layout

### LINKED ISSUE #573: Apply all remaining possible merges after layouting
E.g. iterate over all bricks, sorted by ascending size, look for merges, apply them.

Currently the random function creates a better layout than a deterministic merging would do, but some of the bricks are never selected and therefore never merged, even though there would be mergable neighbors.

### PR DESCRIPTION
**Pull request by johannesdeselaers**
I added an additional pass that handles every brick once
closes #573
There are still some unmerged bricks that could be merged, I believe those are because of #471/#482, which is my next stop

### CODE REVIEW

File: src/plugins/newBrickator/pipeline/BrickLayouter.coffee [hunk 1]
```diff
@@ -52,22 +52,42 @@ class BrickLayouter
...
 					continue # randomly choose a new brick
 
-			while(@_anyDefinedInArray(mergeableNeighbors))
-				mergeIndex = @_chooseNeighborsToMergeWith mergeableNeighbors
-				neighborsToMergeWith = mergeableNeighbors[mergeIndex]
+			@_mergeLoop brick, mergeableNeighbors, bricksToLayout
 
-				@_mergeBricksAndUpdateGraphConnections brick,
-					neighborsToMergeWith, bricksToLayout
+		return {grid: grid}
```
Comment (Steditor) [inline src/plugins/newBrickator/pipeline/BrickLayouter.coffee]: Equal to both `return { grid }` and `return grid: grid`

File: src/plugins/newBrickator/pipeline/BrickLayouter.coffee [hunk 2]
```diff
@@ -52,22 +52,42 @@ class BrickLayouter
...
-					log.warn 'Invalid brick: ', brick
-					log.warn '> Using pseudoRandom:', @pseudoRandom
-					log.warn '> current seed:', Random.getSeed()
 
-				mergeableNeighbors = @_findMergeableNeighbors brick
+	finalLayoutPass: (grid) =>
+		bricksToLayout = grid.getAllBricks()
+		finalPassMerges = 0
+		bricksToLayout.forEach (brick) =>
+			if !brick?
```
Comment (Steditor) [inline src/plugins/newBrickator/pipeline/BrickLayouter.coffee]: `return unless brick?`

Comment (johannesdeselaers) [inline src/plugins/newBrickator/pipeline/BrickLayouter.coffee]: while iterating over the set, some bricks are deleted when they're merged into another brick, it crashed when i removed the check
is there a better way to iterate over all members of a set, while it is being modified?

File: src/plugins/newBrickator/pipeline/BrickLayouter.coffee (hunk 1 above)
Comment (johannesdeselaers) [inline src/plugins/newBrickator/pipeline/BrickLayouter.coffee]: as far as I remember @AtiX introduced this syntax for the multiple legoPipeline passes

File: src/plugins/newBrickator/pipeline/BrickLayouter.coffee (hunk 2 above)
Comment (Steditor) [inline src/plugins/newBrickator/pipeline/BrickLayouter.coffee]: My suggestion is simply a conciser form of your code - `return unless brick?` is equivalent to
```
if !brick?
    return
```

**Comment by AtiX** [comment]
tests: In contrast to dev, the amount of 1x1 plates gets reduced by about 30%. Now, the most used plates are 1x3 plates

**Pull request merged**